<launch>

    <!--Parsing Argument-->

    <!--Proportional Gains-->
    <arg name="kp_x" default="0.1" />
    <arg name="kp_y" default="0.1" />
    <arg name="kp_z" default="0.1" />

    <arg name="kv_x" default="0.05" />
    <arg name="kv_y" default="0.05" />
    <arg name="kv_z" default="0.1" />

    <!--Integral Gains-->
    <arg name="ki_x" default="0.00001" />
    <arg name="ki_y" default="0.00001" />
    <arg name="ki_z" default="0.00001" />

    <arg name="kvi_x" default="0.0000" />
    <arg name="kvi_y" default="0.0000" />
    <arg name="kvi_z" default="0.0000" />

    <!--Derivative Gains-->
    <arg name="kd_x" default="0.0001" />
    <arg name="kd_y" default="0.0001" />
    <arg name="kd_z" default="0.0001" />

    <arg name="kvd_x" default="0.01" />
    <arg name="kvd_y" default="0.01" />
    <arg name="kvd_z" default="0.01" />

    <!--Setpoints-->
    <arg name="xd_StP" default="3.0" />
    <arg name="yd_StP" default="-2.0" />
    <arg name="zd_StP" default="-0.4" />

    <arg name="max_v" default="1.0" />
    <arg name="pitch_bias" default="0.000" />
    <arg name="roll_bias" default="-0.000" />

    <!--Threshold-->
    <arg name="integral_threshold_x" default="10" />
    <arg name="integral_threshold_y" default="10" />
    <arg name="integral_threshold_z" default="10" />

    <arg name="integral_threshold_vx" default="5" />
    <arg name="integral_threshold_vy" default="5" />
    <arg name="integral_threshold_vz" default="5" />

    <!--Running nodes-->
	<node name="distance_collection" pkg="coop_con_loc" type="distance_collection.py" output="screen" />	
	<node name="localization_main" pkg="coop_con_loc" type="localization_main" output="screen" />
	<node name="absolute_velocity_main" pkg="coop_con_loc" type="absolute_velocity_main" output="screen" />
	<node name="relative_Velocity_leader" pkg="coop_con_loc" type="relative_velocity_leader.py" output="screen" />

	<node name="controller" pkg="coop_con_loc" type="controller.py" output="screen" >
        
        <!--Proportional Gains-->
        <param name="kp_x" value="$(arg kp_x)" />
        <param name="kp_y" value="$(arg kp_y)" />
        <param name="kp_z" value="$(arg kp_z)" />

        <param name="kv_x" value="$(arg kv_x)" />
        <param name="kv_y" value="$(arg kv_y)" />
        <param name="kv_z" value="$(arg kv_z)" />

        <!--Integral Gains-->
        <param name="ki_x" value="$(arg ki_x)" />
        <param name="ki_y" value="$(arg ki_y)" />
        <param name="ki_z" value="$(arg ki_z)" />

        <param name="kvi_x" value="$(arg kvi_x)" />
        <param name="kvi_y" value="$(arg kvi_y)" />
	<param name="kvi_z" value="$(arg kvi_z)" />

        <!--Derivative Gains-->
        <param name="kd_x" value="$(arg kd_x)" />
        <param name="kd_y" value="$(arg kd_y)" />
        <param name="kd_z" value="$(arg kd_z)" />

        <param name="kvd_x" value="$(arg kvd_x)" />
        <param name="kvd_y" value="$(arg kvd_y)" />
        <param name="kvd_z" value="$(arg kvd_z)" />

        <!--Setpoints-->
        <param name="xd_StP" value="$(arg xd_StP)" />
        <param name="yd_StP" value="$(arg yd_StP)" />
        <param name="zd_StP" value="$(arg zd_StP)" />

        <param name="max_v" value="$(arg max_v)" />
        <param name="pitch_bias" value="$(arg pitch_bias)" />
        <param name="roll_bias" value="$(arg roll_bias)" />

        <param name="integral_threshold_x" value="$(arg integral_threshold_x)" />
        <param name="integral_threshold_y" value="$(arg integral_threshold_y)" />
        <param name="integral_threshold_z" value="$(arg integral_threshold_z)" />

        <param name="integral_threshold_vx" value="$(arg integral_threshold_vx)" />
        <param name="integral_threshold_vy" value="$(arg integral_threshold_vy)" />
        <param name="integral_threshold_vz" value="$(arg integral_threshold_vz)" />
    </node>

	<node name="px4_communication" pkg="coop_con_loc" type="px4_communication.py" output="screen" />
</launch>
